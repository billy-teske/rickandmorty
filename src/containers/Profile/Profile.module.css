.content {
    display: flex;
    background-size: cover;
    background-attachment: fixed;
}

.content section {
    backdrop-filter: blur(1em);
    background-image: linear-gradient(
            45deg,
            var(--background-color-full),
            #60606030,
            var(--background-color-full)
        ),
        radial-gradient(at top left, var(--background-color-full), #60606030);
    display: grid;
    gap: 2em;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto auto 1fr;
    min-height: 50vh;
    justify-content: center;
    min-height: 100vh;
    padding: 4em 0;
    width: 100%;
}

.content section * {
    animation: fade 0.3s ease;
}

@keyframes fade {
    from {
        opacity: 0;
    }
}

.content img,
.img {
    border-radius: 1em;
    grid-row: 1 / span 2;
    margin: 0 0 0 auto;
    max-width: 60vw;
    background-color: var(--background-color);
    width: 300px;
    height: 300px;
}

.content h2,
.h2 {
    font-size: 2em;
    margin: 0;
    padding: 0 2em;
    text-align: left;
}

.h2 {
    background-color: var(--background-color);
    border-radius: 1em;
    margin-right: auto;
    margin-left: 2em;
    width: 5em;
    height: 1.5em;
}

.content ul,
.ul {
    margin: 0 0 0 4em;
    padding: 0;
}

.content li,
.li {
    margin: 0;
    padding: 0.5em 1em;
    text-align: left;
}

.li {
    background-color: var(--background-color);
    width: 10em;
    border-radius: 1em;
    height: 1em;
    margin: 1em auto 1em 0;
}

.button {
    grid-column: 1 / span 2;
    margin: 0 auto auto auto;
}

@media (max-width: 767px) {
    .content section {
        grid-template-columns: auto;
        grid-template-rows: auto auto auto 1fr;
    }

    .content h2 {
        padding: 0;
        text-align: center;
    }

    .h2 {
        padding: 1em;
        margin: auto;
    }

    .content img,
    .img {
        grid-row: 2;
    }

    .button {
        grid-column: 1;
    }
}
