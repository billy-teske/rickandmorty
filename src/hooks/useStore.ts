import { create } from 'zustand';
import { TCharacter } from '../api/getCharacter';

type TState = {
    characters: TCharacter[];
    setCharacters: (newItems: TCharacter[]) => void;
    page: number;
    nextPage: number;
    toPage: () => void;
    toNextPage: () => void;
};

const useStore = create<TState>((set) => ({
    characters: [],
    setCharacters: (newsCharacters) => set((state) => {
        const addedCharacters = newsCharacters.filter(
            (newCharacter) => !state.characters.find(
                character => newCharacter.id === character.id
            )
        );

        return {
            characters: [...state.characters, ...addedCharacters],
        };
    }),
    page: 0,
    toPage: () => set((state) => ({ page: state.nextPage })),
    nextPage: 1,
    toNextPage: () => set((state) => ({ nextPage: state.page + 1 })),
}));

export default useStore;
