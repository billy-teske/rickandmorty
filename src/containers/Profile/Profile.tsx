import { useEffect } from 'react';
import { Link, useParams } from 'react-router-dom';
import useStore from '../../hooks/useStore';
import styles from './Profile.module.css';

const Profile = () => {
    const { id } = useParams();
    const character = useStore(state => state.characters.find(
        character => !!id && character.id === parseInt(id)
    ));

    useEffect(() => {
        window.scrollTo(0, 0);
    }, []);

    if (!character) {
        return (
            <article className={styles.content}>
                <section>
                    <div className={styles.img}></div>
                    <div className={styles.h2}></div>
                    <div className={styles.ul}>
                        <div className={styles.li}></div>
                        <div className={styles.li}></div>
                        <div className={styles.li}></div>
                        <div className={styles.li}></div>
                    </div>
                    <Link to="/" className={styles.button}>Go Back</Link>
                </section>
            </article>
        );
    }

    const {
        name,
        status,
        species,
        type,
        gender,
        // origin,
        // location,
        image,
        // episode,
    } = character;

    return (
        <article className={styles.content} style={{ backgroundImage: `url(${image})` }}>
            <section>
                <img src={image} alt={`image ${name}`} />
                <h2>{name}</h2>
                <ul>
                    <li>Type {type}</li>
                    <li>Species {species}</li>
                    <li>Gender {gender}</li>
                    <li>Status {status}</li>
                </ul>
                <Link to="/" className={styles.button}>Go Back</Link>
            </section>
        </article>
    );
};

export default Profile;
