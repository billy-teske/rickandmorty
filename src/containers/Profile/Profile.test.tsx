import { render } from '@testing-library/react';
import { Link, useParams } from 'react-router-dom';
import useStore from '../../hooks/useStore';
import Profile from './Profile';
import characterMock from '../../api/__mock__/characterMock';

jest.mock('react-router-dom', () => ({
    Link: jest.fn(),
    useParams: jest.fn(),
}));

jest.mock('../../hooks/useStore', () => jest.fn());

window.scrollTo = jest.fn();
const scrollToMock = jest.spyOn(window, 'scrollTo');

(Link as unknown as jest.Mock).mockImplementation(() => <span>Link</span>);
(useStore as unknown as jest.Mock).mockImplementation(
    (fn) => fn({ characters: characterMock.results })
);

describe('Profile Component', () => {
    it('should render component', () => {
        (useParams as unknown as jest.Mock).mockReturnValue({ id: 361 });

        const result = render(<Profile />);

        expect(result).toMatchSnapshot();
        expect(scrollToMock).toHaveBeenCalledWith(0, 0);
        scrollToMock.mockRestore();
    });

    it('should render skeleton when character no exist', () => {
        (useParams as unknown as jest.Mock).mockReturnValue({ id: 44 });

        const result = render(<Profile />);

        expect(result).toMatchSnapshot();
    });
});
