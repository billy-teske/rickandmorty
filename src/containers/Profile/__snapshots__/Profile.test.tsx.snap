// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="content"
        style="background-image: url(https://rickandmortyapi.com/api/character/avatar/361.jpeg);"
      >
        <section>
          <img
            alt="image Toxic Rick"
            src="https://rickandmortyapi.com/api/character/avatar/361.jpeg"
          />
          <h2>
            Toxic Rick
          </h2>
          <ul>
            <li>
              Type 
              Rick's Toxic Side
            </li>
            <li>
              Species 
              Humanoid
            </li>
            <li>
              Gender 
              Male
            </li>
            <li>
              Status 
              Dead
            </li>
          </ul>
          <span>
            Link
          </span>
        </section>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="content"
      style="background-image: url(https://rickandmortyapi.com/api/character/avatar/361.jpeg);"
    >
      <section>
        <img
          alt="image Toxic Rick"
          src="https://rickandmortyapi.com/api/character/avatar/361.jpeg"
        />
        <h2>
          Toxic Rick
        </h2>
        <ul>
          <li>
            Type 
            Rick's Toxic Side
          </li>
          <li>
            Species 
            Humanoid
          </li>
          <li>
            Gender 
            Male
          </li>
          <li>
            Status 
            Dead
          </li>
        </ul>
        <span>
          Link
        </span>
      </section>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Profile Component should render skeleton when character no exist 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="content"
      >
        <section>
          <div
            class="img"
          />
          <div
            class="h2"
          />
          <div
            class="ul"
          >
            <div
              class="li"
            />
            <div
              class="li"
            />
            <div
              class="li"
            />
            <div
              class="li"
            />
          </div>
          <span>
            Link
          </span>
        </section>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="content"
    >
      <section>
        <div
          class="img"
        />
        <div
          class="h2"
        />
        <div
          class="ul"
        >
          <div
            class="li"
          />
          <div
            class="li"
          />
          <div
            class="li"
          />
          <div
            class="li"
          />
        </div>
        <span>
          Link
        </span>
      </section>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
